### ISSUE TITLE
addRecords if one fails entire operation fails

### ISSUE THREAD

**Issue by osseonews** [gristlabs/grist-api#18]
In running the addRecords, if we tried to add multiple records at once, if one add fails the rest of the adds also won't execute. Sort of works like a promise.all. I think this is the intended functionality, but it is not a good outcome. If we want to add 100 rows, let's say, and 1 row has an issue, the other 99 should get created, and the row with an issue should fail with the proper error reported. However, what happens is that the entire operation fails and no records are created.

**Comment by dsagal** [comment]
I think that depends on the situation, and on the error. E.g. if the failure is a sign of a bug, that it may be much better to fail the whole operation than to let a bunch of corrupted records get created. What kind of a failure are you thinking of (the kind that may affect just one of the records)?

**Comment by osseonews** [comment]
Yeah, there are some failures that are result of a bug, but usually that's not the case when inserting multiple records. It's actually an edge case, and in the vast majority of cases, that is not what you want.  For example, you may have an access control rule that prevents adding records when there is a duplicate email for a customer. If then let's say we try to add 20 records, and 1 record is a duplicate, then why should the other 19 inserts fail when there is nothing wrong with them? The better way to do it is to have the other 19 records insert and just have an error for that 1 problem insert. And also, if there is a bug, all the inserts would fail anyway, so that scenario would also be handled fine. Overall, I don't see any benefit of having the whole transaction fail ever. Plus, if you want that kind of functionality you can just create a promise.all wrapper around the api and achieve the same effect.

**Comment by dsagal** [comment]
The scenario with preventing duplicates makes sense.

It's not an easy fix though: there is good reason to add all records in a single API call -- for efficiency, and to avoid API call limits. A batch call, however, translates to a batch action internally in Grist. Access Rules (like the one to prevent duplicates in your example) act on the entire action: the action is either in or out. So in order to implement this kind of granular action, Grist would need to add support for something like an [ON CONFLICT](https://www.postgresql.org/docs/current/sql-insert.html#SQL-ON-CONFLICT) clause in SQL.

This makes sense to me, in particular because (specifically for duplicates at least) it mostly matches functionality that already exists for imports, but isn't exposed in API. It would need to be implemented in https://github.com/gristlabs/grist-core first: I suggest filing an issue there (and please include your example if you do, it is definitely helpful).

Meanwhile, I see two workarounds. The one I do NOT recommend is to make 19 separate calls, one for each record. The workaround I would use is to fetch the entire table first, use it to prune duplicates from the list of emails I am adding, and then make a single batch `addRecords` call with just the records that I know should succeed.
